deployment/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,392 @@
+	Stage  string         `json:"stage"`
+	Index  indexResponse  `json:"index"`
+	Source sourceResponse `json:"source"`
+}
+
+type indexRecoveryResponse struct {
+	Shards []shardResponse `json:"shards"`
+}
+
+type indicesRecoveryResponse map[string]indexRecoveryResponse
+
+// IndexShardRecovery represents the information returned by the IndicesRecovery
+// request for a single index shard, specifying:
+//   - Index: the name of the index where this shard lives.
+//   - Type: the type of the shard, normally "SNAPSHOT" for shards that are
+//     being restored from a snapshot.
+//   - Stage: the current status of the shard. For SNAPSHOT shards, this is
+//     either "INDEX", meaning the shard is still being restored, or "DONE",
+//     meaning the shard has already been restored.
+//   - Percent: the percentage of bytes already restored. Note this is not the
+//     percentage of files already restored.
+type IndexShardRecovery struct {
+	Index   string
+	Type    string
+	Stage   string
+	Percent string
+}
+
+// IndicesRecovery returns status information for each index shard in the server.
+// This is useful to track the completion of a snapshot restoration process.
+func (c *Client) IndicesRecovery(indices []string) ([]IndexShardRecovery, error) {
+	req := esapi.IndicesRecoveryRequest{
+		Index: indices,
+	}
+	indicesRecovery := make(indicesRecoveryResponse)
+	if err := c.get(req, &indicesRecovery); err != nil {
+		return nil, fmt.Errorf("unable to perform IndicesRecovery request: %w", err)
+	}
+
+	recovery := []IndexShardRecovery{}
+	for _, resp := range indicesRecovery {
+		for _, shard := range resp.Shards {
+			recovery = append(recovery, IndexShardRecovery{
+				Index: shard.Source.Index,
+				Type:  shard.Type,
+				Stage: shard.Stage,
+				// We're using the percentage of bytes restored,
+				// not the percentage of files restored
+				Percent: shard.Index.Size.Percent,
+			})
+		}
+	}
+
+	return recovery, nil
+}
+
+// requestDoer models all esapi.XYZRequest, which contains a Do function to
+// perform the request with the provided client
+type requestDoer interface {
+	Do(context.Context, esapi.Transport) (*esapi.Response, error)
+}
+
+// get runs req.Do, performs the needed checks on the response, and stores the
+// result in the value pointed to by result
+func (c *Client) get(req requestDoer, result any) error {
+	res, err := req.Do(context.Background(), c.client)
+	if err != nil {
+		return fmt.Errorf("unable to perform request: %w", err)
+	}
+	if res.Body == nil {
+		return fmt.Errorf("no body returned")
+	}
+	defer res.Body.Close()
+	if res.IsError() {
+		// Consume body, docs say it's important to do so even if not needed
+		io.Copy(io.Discard, res.Body)
+		return fmt.Errorf("request failed: %q", res.String())
+	}
+
+	resBytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(resBytes, result); err != nil {
+		return fmt.Errorf("unable to unmarshal response: %w", err)
+	}
```

Comment:
Not a big deal but we could use a `json.Decoder` instead for more optimal usage.

Reply:
Ah, yes, always forget about them!